gcc/rust/typecheck/rust-tyty-variance-analysis.cc

```diff
@@ -0,0 +1,540 @@
+#include "rust-tyty-variance-analysis-private.h"
+#include "rust-hir-type-check.h"
+
+namespace Rust {
+namespace TyTy {
+
+BaseType *
+lookup_type (HirId ref)
+{
+  BaseType *ty = nullptr;
+  bool ok = Resolver::TypeCheckContext::get ()->lookup_type (ref, &ty);
+  rust_assert (ok);
+  return ty;
+}
+
+namespace VarianceAnalysis {
+
+void
+add_type_constraints (ADTType &type)
+{
+  GenericTyVisitorCtx::get ().process_type (type);
+}
+
+void
+solve ()
+{
+  auto ctx = GenericTyVisitorCtx::get ();
```

Comment:
I'm not a big fan of adding more global context to the frontend, but if it can't be helped them I'm also fine with it. note that this might cause hurt down the line with the compilation of multiple crates, proc macros, possible refactors for a more re-entrant frontend, etc

Reply:
I guess, we could tie it to the typecheck context and pass a pointer to this method. Does that look good to you?